adapters/src/main/kotlin/com/krzykrucz/elesson/currentlesson/preparedness/adapters/rest/UnpreparedStudentHttpEndpoint.kt

```diff
@@ -0,0 +1,71 @@
+package com.krzykrucz.elesson.currentlesson.preparedness.adapters.rest
+
+import arrow.core.Tuple2
+import arrow.core.getOrElse
+import arrow.core.getOrHandle
+import arrow.core.right
+import arrow.effects.typeclasses.Duration
+import com.krzykrucz.elesson.currentlesson.MonoDomainError
+import com.krzykrucz.elesson.currentlesson.handleErrors
+import com.krzykrucz.elesson.currentlesson.preparedness.domain.FindCurrentLesson
+import com.krzykrucz.elesson.currentlesson.preparedness.domain.NotifyStudentMarkedUnprepared
+import com.krzykrucz.elesson.currentlesson.preparedness.domain.PersistUnpreparedStudentToLesson
+import com.krzykrucz.elesson.currentlesson.preparedness.domain.ReportUnpreparedness
+import com.krzykrucz.elesson.currentlesson.preparedness.domain.StudentReportingUnpreparedness
+import com.krzykrucz.elesson.currentlesson.preparedness.readmodel.WriteUnpreparednessInTheRegister
+import com.krzykrucz.elesson.currentlesson.shared.LessonIdentifier
+import com.krzykrucz.elesson.currentlesson.shared.flatMapAsyncSuccess
+import com.krzykrucz.elesson.currentlesson.shared.mapSuccess
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+import org.springframework.web.reactive.function.server.ServerResponse
+import org.springframework.web.reactive.function.server.router
+import reactor.core.publisher.Mono
+import java.util.concurrent.TimeUnit
+
+data class ReportUnpreparedRequest(
+    val studentName: StudentReportingUnpreparedness,
+    val lessonIdentifier: LessonIdentifier
+)
+
+
+@Configuration
+class ReportUnpreparedRouteAdapter {
+
+    @Bean
+    fun reportUnpreparedRoute(
+        reportUnpreparedness: ReportUnpreparedness,
+        writeUnpreparednessInTheRegister: WriteUnpreparednessInTheRegister,
+        findCurrentLesson: FindCurrentLesson,
+        persistUnpreparedStudentToLesson: PersistUnpreparedStudentToLesson,
+        notifyStudentMarkedUnprepared: NotifyStudentMarkedUnprepared
+    ) =
+        router {
+            POST("/reportunprepared") { request ->
+                request.bodyToMono(ReportUnpreparedRequest::class.java)
+                    .flatMap { req ->
+                        findCurrentLesson(req.lessonIdentifier)
+                            .mapSuccess { Tuple2(it, req.studentName) }
+                            .flatMapAsyncSuccess { reportUnpreparedness(it.b, it.a) }
+                            .flatMapAsyncSuccess { event ->
+                                persistUnpreparedStudentToLesson(event)
```

Comment:
I doubt if it should be responsibility of an adapter to `persists` or `notifies`, still have this impression that `use-case` layer might be helpful here to encapsulate necessarty orchestration regardless from which adapter request comes from. Or, but here I'm stepping into unknown ground, an implementation of workflow could do it bearing in mind that `persistence` and `notification` functions are defined in the domain, what bothers me also.

Reply:
maybe it would be ok to put all these use-case (application) logic in the domain package? Because we now have 2 layers: domain and adapters. If we created a use-case (application) layer, we will have 3, which might be an overcomplication for a workshop's sake, don't you think? If not, we can refactor the whole project towards 3 layers.

Reply:
it's easier to just put it into domain for now, then we can review how it smells and decide what next

Reply:
+1 for `use-case` layer. In my opinion, having that orchestration layer makes whole architecture clearer. It also enable us to have framework-specific `Mono` layer only in adapters.